File: NetFramework/TreeView/VB/CustomStateImages/ImageEditor.vb
Imports System.Windows.Forms.VisualStyles

Public Class ImageEditor

    Public Sub New(imgList__1 As ImageList)
        InitializeComponent()
        ImgList = imgList__1
        lvImages.StateImageList = New ImageList()
        FillCombobox(cmbImageTypes)
        cmbImageTypes.SelectedIndex = 0
        FillImages()
    End Sub

    Private Sub btnAdd_Click(sender As Object, e As EventArgs) Handles btnAdd.Click
        Dim ofd = New OpenFileDialog()
        ofd.CheckFileExists = True
        ofd.Multiselect = False
        ofd.Filter = "All image files (*.bmp;*.gif;*.jpg;*.jpeg;*.png;*.ico)|*.bmp;*.gif;*.jpg;*.jpeg;*.png;*.ico"
        If ofd.ShowDialog() = DialogResult.OK Then
            AddItem(Image.FromFile(ofd.FileName), cmbImageTypes.SelectedItem.ToString())
        End If
    End Sub

    Private Sub btnRemove_Click(sender As Object, e As EventArgs) Handles btnRemove.Click
        If lvImages.SelectedItems.Count > 0 Then
            RemoveItem(lvImages.SelectedItems(0).Name)
        End If
    End Sub

    Private Sub FillImages()
        For Each key As String In ImgList.Images.Keys
            AddItem(ImgList.Images(key), key)
        Next
    End Sub

    Protected Sub AddItem(image As Image, name As String)
        RemoveItem(name)

        lvImages.StateImageList.Images.Add(name, image)

        Dim item = New ListViewItem()
        item.StateImageIndex = lvImages.StateImageList.Images.IndexOfKey(name)
        item.Name = name
        item.Text = name
        lvImages.Items.Add(item)
    End Sub

    Private Sub RemoveItem(name As String)
        If lvImages.StateImageList.Images.ContainsKey(name) Then
            lvImages.StateImageList.Images.RemoveByKey(name)
        End If

        If lvImages.Items.ContainsKey(name) Then
            lvImages.Items.RemoveByKey(name)
        End If

        For Each item As ListViewItem In lvImages.Items
            item.StateImageIndex = lvImages.StateImageList.Images.IndexOfKey(item.Name)
        Next
    End Sub

    Protected Overridable Sub FillCombobox(cmb As ComboBox)

    End Sub

    Protected Property ImgList() As ImageList
        Get
            Return m_ImgList
        End Get
        Set
            m_ImgList = Value
        End Set
    End Property

    Private m_ImgList As ImageList

    Public Sub FillImageList()
        ImgList.Images.Clear()
        For Each item As ListViewItem In lvImages.Items
            ImgList.Images.Add(item.Name, lvImages.StateImageList.Images(item.StateImageIndex))
        Next
    End Sub

End Class

Public Class CheckStateImageEditor
    Inherits ImageEditor
    Public Sub New(imgList As ImageList)

        MyBase.New(imgList)
    End Sub

    Protected Overrides Sub FillCombobox(cmb As ComboBox)
        For Each key As String In [Enum].GetNames(GetType(CheckBoxState))
            If ImgList.Images.ContainsKey(key) Then
                AddItem(ImgList.Images(key), key)
            End If
            cmb.Items.Add(key)
        Next
    End Sub
End Class

Public Class ButtonStateImageEditor
    Inherits ImageEditor
    Public Sub New(imgList As ImageList)

        MyBase.New(imgList)
    End Sub

    Protected Overrides Sub FillCombobox(cmb As ComboBox)
        AddCmbItem(cmb, "CollapsedNormal")
        AddCmbItem(cmb, "CollapsedHot")
        AddCmbItem(cmb, "CollapsedPressed")
        AddCmbItem(cmb, "CollapsedDisabled")
        AddCmbItem(cmb, "ExpandedNormal")
        AddCmbItem(cmb, "ExpandedHot")
        AddCmbItem(cmb, "ExpandedPressed")
        AddCmbItem(cmb, "ExpandedDisabled")
    End Sub

    Private Sub AddCmbItem(cmb As ComboBox, key As String)
        If ImgList.Images.ContainsKey(key) Then
            AddItem(ImgList.Images(key), key)
        End If
        cmb.Items.Add(key)
    End Sub
End Class

File: NetFramework/TreeView/CS/CustomStateImages/ImageEditor.cs
using C1.Win.Ribbon;
using System;
using System.Drawing;
using System.Windows.Forms;
using System.Windows.Forms.VisualStyles;

namespace CustomStateImages
{
    public partial class ImageEditor : C1RibbonForm
    {
        public ImageEditor(ImageList imgList)
        {
            InitializeComponent();
            ImgList = imgList;
            lvImages.StateImageList = new ImageList();
            FillCombobox(cmbImageTypes);
            cmbImageTypes.SelectedIndex = 0;
            FillImages();
        }

        private void btnAdd_Click(object sender, EventArgs e)
        {
            var ofd = new OpenFileDialog();
            ofd.CheckFileExists = true;
            ofd.Multiselect = false;
            ofd.Filter = "All image files (*.bmp;*.gif;*.jpg;*.jpeg;*.png;*.ico)|*.bmp;*.gif;*.jpg;*.jpeg;*.png;*.ico";
            if (ofd.ShowDialog() == DialogResult.OK)
            {
                AddItem(Image.FromFile(ofd.FileName), cmbImageTypes.SelectedItem.ToString());
            }
        }

        private void btnRemove_Click(object sender, EventArgs e)
        {
            if (lvImages.SelectedItems.Count > 0)
                RemoveItem(lvImages.SelectedItems[0].Name);
        }

        private void FillImages()
        {
            foreach(var key in ImgList.Images.Keys)
            {
                AddItem(ImgList.Images[key], key);
            }
        }

        protected void AddItem(Image image, string name)
        {
            RemoveItem(name);

            lvImages.StateImageList.Images.Add(name, image);

            var item = new ListViewItem();
            item.StateImageIndex = lvImages.StateImageList.Images.IndexOfKey(name);
            item.Name = name;
            item.Text = name;
            lvImages.Items.Add(item);
        }

        private void RemoveItem(string name)
        {
            if (lvImages.StateImageList.Images.ContainsKey(name))
                lvImages.StateImageList.Images.RemoveByKey(name);

            if (lvImages.Items.ContainsKey(name))
                lvImages.Items.RemoveByKey(name);

            foreach (ListViewItem item in lvImages.Items)
                item.StateImageIndex = lvImages.StateImageList.Images.IndexOfKey(item.Name);
        }

        protected virtual void FillCombobox(ComboBox cmb)
        {

        }

        protected ImageList ImgList
        {
            get;
            set;
        }

        public void FillImageList()
        {
            ImgList.Images.Clear();
            foreach(ListViewItem item in lvImages.Items)
                ImgList.Images.Add(item.Name, lvImages.StateImageList.Images[item.StateImageIndex]);            
        }
    }

    public class CheckStateImageEditor : ImageEditor
    {
        public CheckStateImageEditor(ImageList imgList) : base(imgList)
        {

        }

        protected override void FillCombobox(ComboBox cmb)
        {
            foreach (string key in Enum.GetNames(typeof(CheckBoxState)))
            {
                if (ImgList.Images.ContainsKey(key))
                    AddItem(ImgList.Images[key], key);
                cmb.Items.Add(key);
            }            
        }
    }

    public class ButtonStateImageEditor : ImageEditor
    {
        public ButtonStateImageEditor(ImageList imgList) : base(imgList)
        {

        }

        protected override void FillCombobox(ComboBox cmb)
        {
            AddCmbItem(cmb, "CollapsedNormal");
            AddCmbItem(cmb, "CollapsedHot");
            AddCmbItem(cmb, "CollapsedPressed");
            AddCmbItem(cmb, "CollapsedDisabled");
            AddCmbItem(cmb, "ExpandedNormal");
            AddCmbItem(cmb, "ExpandedHot");
            AddCmbItem(cmb, "ExpandedPressed");
            AddCmbItem(cmb, "ExpandedDisabled");
        }

        private void AddCmbItem(ComboBox cmb, string key)
        {
            if (ImgList.Images.ContainsKey(key))
                AddItem(ImgList.Images[key], key);
            cmb.Items.Add(key);
        }
    }
}
